README.md
[![ci-cd](https://github.com/UBC-MDS/fxtracker/actions/workflows/ci-cd.yml/badge.svg?branch=main)](https://github.com/UBC-MDS/fxtracker/actions/workflows/ci-cd.yml) [![codecov](https://codecov.io/gh/UBC-MDS/fxtracker/branch/main/graph/badge.svg?token=N4sBOXKB87)](https://codecov.io/gh/UBC-MDS/fxtracker) [![Documentation Status](https://readthedocs.org/projects/fxtracker/badge/?version=latest)](https://fxtracker.readthedocs.io/en/latest/?badge=latest)

# fxtracker

This is a package created as a group project for DSCI_524 Collaborative Software Development of UBC Master of Data Science (MDS) program 2022-2023. Based on the foreign exchange data in Yahoo Finance, this package allows user to perform currency conversion based on the latest available exchange rate, lookup a target exchange rate from historical data as well plotting exchange rate history and profit/loss percentage history by specifying a currency pair (and other input parameters).

The full documentation of this package can also be found on <https://fxtracker.readthedocs.io/en/latest/>

## Function Description

-   `fx_conversion` <br> Convert the input amount of currency 1 to currency 2 based on the latest available exchange rate.
-   `fx_rate_lookup` <br> Lookup for the most recent date on which the input target rate of a currency pair is within the day's high/low.
-   `price_trend_viz` <br> Plot the historical exchange rate of the input currency pair for a specific period of time.
-   `pl_trend_viz` <br> Plot the historical profit/loss percentage of the input currency pair for a specific period of time.

There is a python package ([`forex-python`](https://pypi.org/project/forex-python/)) relevant to foreign exchange. That package is basically for retrieving exchange rates and bitcoin prices in plain text as well as performing conversion. It does not provide visualizations and lookup function like `fxtracker` does. `fxtracker` allows user to visualize the trends and understand if a target price of a currency pair of interest is within a reasonable range.

## Installation

```bash
$ pip install fxtracker
```

## Usage

If the package is installed successfully, users then need the following nine input parameters:

`curr`, `target_px`, `start_date`, `end_date`, `chart_type`, `option`, `curr1`, `curr2`, `amt`. The output of the functions will be in forms of a datetime string, a float and interactive plots from the "altair" package.

`fxtracker` can be used to convert a specific amount of money from one currency to another, find the most recent date on which the target price falling between day high and day low of that day, visualize the trend of the exchange rate of a currency pair and the trend of the profit and loss of a currency pair between the selected start date and end date.

The functions can be imported from the package as follows:

```python
from fxtracker.fxtracker import fx_conversion
from fxtracker.fxtracker import fx_rate_lookup
from fxtracker.fxtracker import price_trend_viz
from fxtracker.fxtracker import pl_trend_viz
```

### To convert a specific amount of money from current currency (curr1) to desired currency (curr2):

    fx_conversion('EUR', 'USD', 150.75)

163.68

### To look up the most recent date on which the target price falling between day high and day low of that day:

    fx_rate_lookup('EURUSD', 1.072)

'2023-01-10'

### To visualize the trend of the exchange rate of a currency pair between the selected start date and end date:

    price_trend_viz('EURUSD', '2018-12-01', '2022-12-01', 'High').show()

![](https://user-images.githubusercontent.com/112665905/215251534-3d452198-23bc-4b42-885c-d76a5ca68f25.png)

### To visualize the trend of the profit and loss of a currency pair between the selected start date and end date:

**If a line chart is specified in the input:**

    pl_trend_viz("EURUSD", "2020-01-01", "2022-01-01", 'line').show()

![](https://user-images.githubusercontent.com/112665905/215251530-8a3cf86f-6854-47b5-b7b4-2ff214e88217.png)

**If an area chart is specified in the input:**

    pl_trend_viz("EURUSD", "2020-01-01", "2022-01-01", 'area').show()

![](https://user-images.githubusercontent.com/112665905/215251527-3381d5de-c776-4b5f-9777-c687b287f089.png)

## Dependencies

-   python = "\^3.9"
-   pandas = "\^1.5.2"
-   altair = "\^4.2.0"
-   numpy = "\^1.24.1"
-   plotly = "\^5.12.0"
-   yfinance = "\^0.2.3"
-   ipykernel = "\^6.20.2"
-   altair-viewer = "\^0.4.0"

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`fxtracker` was created by Sarah Abdelazim, Markus Nam, Crystal Geng and Lennon Au-Yeung. It is licensed under the terms of the MIT license.

## Credits

`fxtracker` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).

```python

```


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.3 (2023-02-12)
### Documentation
* Update CONTRIBUTING.md ([`d0b6152`](https://github.com/UBC-MDS/fxtracker/commit/d0b615264919e327f872b6226de162a9bbf0205c))

## v1.0.2 (2023-02-12)
### Documentation
* Update comments of fx_rate_lookup() ([`4e0dd32`](https://github.com/UBC-MDS/fxtracker/commit/4e0dd32f936d606c31f812b0a341befea9dd0839))
* Add links to github accounts in CONTRIBUTING.md ([`95d0e8f`](https://github.com/UBC-MDS/fxtracker/commit/95d0e8fa1ad7b501140a026aa07104f2a3107f52))
* Fix a typo in a comment of fx_rate_lookup ([`fd69f46`](https://github.com/UBC-MDS/fxtracker/commit/fd69f4650964a40ca8dcb9e19a7d63fa6b24e710))
* Update README ([`d81ab26`](https://github.com/UBC-MDS/fxtracker/commit/d81ab266b06ee9c674944a873ac046b2c13a1ee0))
* Update fx_rate_lookup comments ([`47ea868`](https://github.com/UBC-MDS/fxtracker/commit/47ea86800bcc2d78b7d3507cebb84661e0d234db))

## v1.0.1 (2023-02-05)


## v1.0.0 (2023-01-29)
### Feature
* New changes ([`5b4922e`](https://github.com/UBC-MDS/fxtracker/commit/5b4922e134a460db2917db7c1438218ffd44c90d))

### Breaking
* new changes ([`5b4922e`](https://github.com/UBC-MDS/fxtracker/commit/5b4922e134a460db2917db7c1438218ffd44c90d))

## v0.2.1 (2023-01-29)


## v0.2.0 (2023-01-29)
### Feature
* **new:** Updated project. BREAKING CHANGE ([`f74ba12`](https://github.com/UBC-MDS/fxtracker/commit/f74ba12ed03bd4688be154d28b27a7e930c79889))

## v0.1.1 (2023-01-29)


## v0.1.0 (2023-01-29)
### Feature
* Add input check for fx_rate_lookup() ([`4453713`](https://github.com/UBC-MDS/fxtracker/commit/445371316d55a607b1d9a89df03fc1f723c9e3fe))
* Implement fx_rate_lookup() ([`f96fb43`](https://github.com/UBC-MDS/fxtracker/commit/f96fb43e316d5b21956dc22eea12c5c96cb8006d))
* Add function definition for fx_rate_lookup ([`d1d5b17`](https://github.com/UBC-MDS/fxtracker/commit/d1d5b176331ad13b003b11c41f6ac26fbaa929de))

### Fix
* Fix cd ([`99a0845`](https://github.com/UBC-MDS/fxtracker/commit/99a0845bf31aeba45b92d3b4d7f79542ce4ac9f8))
* Fix cd ([`85aadf6`](https://github.com/UBC-MDS/fxtracker/commit/85aadf6f809be6a8e6d9e73a4cde942c36f3d2dd))
* Fix cd ([`311016e`](https://github.com/UBC-MDS/fxtracker/commit/311016e85492a2deb105adff0455ff2e8b45f6b0))
* Only trigger ci-cd when push or PR to main ([`f8473c1`](https://github.com/UBC-MDS/fxtracker/commit/f8473c1911553bee094d690a713b7c93ef442361))
* Fix the return string of fx_rate_lookup ([`4043944`](https://github.com/UBC-MDS/fxtracker/commit/40439445d4f09159bf6dd31a3cbace538c279996))
* Fix return string of fx_rate_lookup() ([`63a310d`](https://github.com/UBC-MDS/fxtracker/commit/63a310d5f10df5735a3c57b1377751a877a55080))

### Documentation
* Update README.md ([`6bbf196`](https://github.com/UBC-MDS/fxtracker/commit/6bbf19640a36e7fd4e1e97c66fbb766ba789f3d4))

## v0.1.0 (12/01/2023)

- First release of `fxtracker`!

tests/test_fxtracker.py
from fxtracker.fxtracker import fx_rate_lookup
from fxtracker.fxtracker import pl_trend_viz
from fxtracker.fxtracker import price_trend_viz
from fxtracker.fxtracker import fx_conversion

import pandas as pd
import datetime
import pytest
from pytest import raises
import altair as alt
import numpy as np


def test_fx_rate_lookup():

    with raises(TypeError):
        fx_rate_lookup(123, 456)

    with raises(TypeError):
        fx_rate_lookup('JPYHKD', "456")

    with pytest.raises(Exception):
        fx_rate_lookup('AAABBB', 0.09)

    with pytest.raises(Exception):
        fx_rate_lookup('JPYHKD', 1)

    with pytest.raises(Exception):
        fx_rate_lookup()

    with pytest.raises(Exception):
        fx_rate_lookup('JPYHKD')

    assert isinstance(pd.to_datetime(fx_rate_lookup('JPYHKD', 0.09)),
                      datetime.datetime), (
        'Return string should be a valid YYYY-MM-DD format')


def test_pl_trend_viz():

    # Curr type invalid
    with raises(TypeError):
        pl_trend_viz(1, "20200101", "2022-01-10", 'line')

    # Start date type invalid
    with raises(TypeError):
        pl_trend_viz("EURUSD", 1234, "20220110", 'line')

    # End date type invalid
    with raises(TypeError):
        pl_trend_viz("EURUSD", "2020-01-01", 1234, 'line')

    # Type of visulization error
    with raises(ValueError):
        pl_trend_viz("EURUSD", "2020-01-01", "2022-01-01", 'point')

    # Start date > end date
    with raises(ValueError):
        pl_trend_viz("EURUSD", "2022-01-01", "2020-01-10", 'line')

    # Curr name error
    with raises(NameError):
        pl_trend_viz("ABCD", "2020-01-01", "2022-01-01", 'line')

    output = pl_trend_viz("EURUSD", "2020-01-01", "2022-01-01", 'area')
    assert isinstance(
        output,
        alt.vegalite.v4.api.Chart), "Altair chart object should be returned."

    output2 = pl_trend_viz("EURUSD", "2020-01-01", "2022-01-01", 'line')
    assert isinstance(
        output2,
        alt.vegalite.v4.api.Chart), "Altair chart object should be returned."

    raw = output.to_dict()

    assert raw['encoding']['x']['field'] == 'Date', (
            'Date should be mapped to the x axis')

    assert raw['encoding']['y']['field'] == 'Percentage Change', (
            'Percentage Change should be mapped to the y axis')


def test_price_trend_viz():

    # Check input type of curr
    with pytest.raises(TypeError, match="curr needs to be of str type."):
        price_trend_viz(1, '2012-02-01', '2014-12-31', 'High')

    # Check input type of start_date
    with pytest.raises(TypeError, match="start_date needs to be of str type."):
        price_trend_viz('EURUSD', 2.5, '2014-12-31', 'Low')

    # Check input type of end_date
    with pytest.raises(TypeError, match="end_date needs to be of str type."):
        price_trend_viz('EURUSD', '2012-02-01', 333, 'Close')

    with pytest.raises(Exception, match="Your option of plotting"
                       " should be from 'Open', 'High', 'Low' or 'Close'"):
        price_trend_viz('EURUSD', '2012-12-31', '2014-12-31', 'Highest')

    with pytest.raises(Exception, match="Your option of plotting"
                       " should be from 'Open', 'High', 'Low' or 'Close'"):
        price_trend_viz('EURUSD', '2012-12-31', '2014-12-31', 111)

    # Check if the end date entered is later than 2003-12-01
    with pytest.raises(Exception, match="No data exists before"
                       " 2003-12-01, please try again."):
        price_trend_viz('EURUSD', '1992-02-01', '1994-12-31', 'High')

    # Check if the start date entered is earlier than or equal to today
    with pytest.raises(Exception, match="You entered a start date later"
                       " than today, please try again."):
        price_trend_viz('EURUSD', '2023-12-31', '2024-12-31', 'High')

    # Check if the same dates are entered for start date and end date
    with pytest.raises(Exception, match="No data found from data source."
                       " Check your ticker and date."):
        price_trend_viz('EURUSD', '2013-12-31', '2013-12-31', 'High')

    output_chart = price_trend_viz(
        'EURUSD', '2018-12-01', '2022-12-01', 'High')

    # Check if an Altair chart is returned, code referenced from pystockwatch
    assert isinstance(
        output_chart,
        alt.vegalite.v4.api.Chart), "Altair chart object should be returned."


def test_fx_conversion():

    # Check if a numpy float is returned
    assert isinstance(fx_conversion('EUR', 'USD', 100),
                      np.float64), "Float should be returned."

    # Check input type of curr_1
    with pytest.raises(TypeError, match="Ticker of the current"
                       " currency needs to be of str type."):
        fx_conversion(1, 'USD', 100)

    # Check input type of curr_2
    with pytest.raises(TypeError, match="Ticker of the desired "
                       "currency needs to be of str type."):
        fx_conversion('EUR', 1, 100)

    # Check input type of amt
    with pytest.raises(TypeError, match="Amount of money to be "
                       "converted needs to be a number."):
        fx_conversion('EUR', 'USD', 'a')

    # Check if no arguments are passed
    with pytest.raises(TypeError,
                       match=r"^fx_conversion\(\) missing 3 required"
                       " positional arguments: 'curr1', 'curr2', and 'amt'$"):
        fx_conversion()

    # Check if curr_2 and amt arguments are not passed
    with pytest.raises(TypeError, match=r"^fx_conversion\(\) missing 2 "
                       "required positional arguments: 'curr2' and 'amt'$"):
        fx_conversion('USD')

    # Check if amt argument is not passed
    with pytest.raises(TypeError, match=r"^fx_conversion\(\) missing 1 "
                       "required positional argument: 'amt'$"):
        fx_conversion('USD', 'EUR')

    # Check if wrong ticker(s) is/are passed
    with pytest.raises(NameError,
                       match="You have entered an "
                       "invalid foreign ticker! Try again."):
        fx_conversion('XXX', 'YYY', 100)
    with pytest.raises(NameError,
                       match="You have entered an "
                       "invalid foreign ticker! Try again."):
        fx_conversion('USD', 'YYY', 100)
    with pytest.raises(NameError,
                       match="You have entered an "
                       "invalid foreign ticker! Try again."):
        fx_conversion('XXX', 'EUR', 100)


tests/test_pl_trend_viz.py
from fxtracker.pl_trend_viz import *

import altair as alt
import pytest
from pytest import raises
import pandas as pd

def test_pl_trend_viz():
    
    # Invalid input for start date
    with raises(ValueError) as error_start_date:
        pl_trend_viz("EURUSD", "20200101", "2022-01-10")

    # Invalid input for end date
    with raises(ValueError) as error_end_date:
        pl_trend_viz("EURUSD", "2020-01-01", "20220110")

    # End date is ealier than start date
    with raises(ValueError) as error_end_date:
        pl_trend_viz("EURUSD", "2022-01-01", "2020-01-10")
        
    output = pl_trend_viz("EURUSD", "2020-01-01", "2022-01-01")
    assert type(output) == alt.vegalite.v4.api.Chart, "Altair chart object should be returned."

src/fxtracker/pl_trend_viz.py
# Authors: Sarah Abdelazim, Markus Nam, Crystal Geng, Lennon Au-Yeung
# Date: Jan 13, 2023

import datetime
import pandas as pd
import altair as alt
import numpy as np
import yfinance as yf
from yfinance import shared
import altair_viewer
alt.renderers.enable('mimetype')

def pl_trend_viz(curr, start_date, end_date, chart_type):
    """
    Visualizes trend of the profit and loss of a currency pair between the selected start date and end date.
    
    Parameters
    ----------
    curr : string 
        Ticker of the currency pair such as 'EURUSD'
    start_date : string 
        Start date of the selected period of time
    end_date : string
        End date of the selected period of time
    chart_type: string
        Type of visulization
    
    Returns
    --------
    Line plot that shows the trend of the profit and loss of a currency pair over the selected period of time
    
    Examples
    --------
    >>> pl_trend_viz('EURUSD', '2018-12-31', '2022-12-31','line')
    """
    shared._ERRORS = {}
    curr_X = curr +'=X'
    
    #check input type of curr
    if not isinstance(curr, str):
        raise TypeError("curr needs to be of str type.")
        
    # Check input type of start_date     
    if not isinstance(start_date, str):
        raise TypeError("start_date needs to be of str type.")
        
    # Check input type of end_date  
    if not isinstance(end_date, str):
        raise TypeError("end_date needs to be of str type.")
    
    #Check type of visulization
    if chart_type not in ['line', 'area']:
        raise ValueError("Your option of plotting should be from 'line' or 'area'")

    # Assert end date is later than start date
    format = "%Y-%m-%d"
    if(datetime.datetime.strptime(end_date, format) < datetime.datetime.strptime(start_date, format)):
        raise ValueError("You have entered an end date which is earlier than the start date! Try again.")
    
    data = yf.download(curr_X, start_date, end_date)
    
    if curr_X in shared._ERRORS and 'symbol may be delisted' in shared._ERRORS[curr_X] :
        raise NameError("You have entered an invalid foreign ticker! Try again.")
    
    drop = {'Open', 'High', 'Low', 'Adj Close', 'Volume'}
    data = data.drop(columns = drop)
    
    for i in range(1,len(data)):
        data.iloc[i,:] = round((data.iloc[i,:] - data.iloc[0,:])/data.iloc[0,:], 3)
    
    data.iloc[0,:] = round((data.iloc[0,:] - data.iloc[0,:])/data.iloc[0,:], 3)
    
    data = data.rename(columns={"Close": "Percentage Change"})
    
    title = 'Price trend over time for currency ' + curr
    
    if chart_type == 'line':
        chart = alt.Chart(data.reset_index(),title = title).mark_line().encode(
            x = alt.X('Date'),
            y = alt.Y('Percentage Change',axis=alt.Axis(format='%')),
            tooltip = alt.Tooltip('Percentage Change',format = '.2%')
        )
    else:
        chart = alt.Chart(data.reset_index(),title = title).mark_area().encode(
            x = alt.X('Date'),
            y = alt.Y('Percentage Change',axis=alt.Axis(format='%')),
            tooltip = alt.Tooltip('Percentage Change',format = '.2%')
        )
    
    chart.show()
    
    return chart

src/fxtracker/fx_conversion.py
import pandas as pd
import numpy as np
import yfinance as yf
from yfinance import shared

def fx_conversion(curr1, curr2, amt):
    """
    Converts a specific amount of money from current currency (curr1) to desired currency (curr2)
    
    Parameters
    ----------
    curr1 : string 
        Ticker of the current currency such as 'EUR'
    curr2 : string 
        Ticker of the desired currency to convert to such as 'USD'
    amt : int or float
        amount of money to be converted

    
    Returns
    --------
    Amount of value converted from currency 1 to currency 2
    
    Examples
    --------
    >>> fx_conversion('EUR', 'USD', 150.75)
    """
    shared._ERRORS = {}
    ##Tests
    # Check input type of curr1
    if not isinstance(curr1, str):
        raise TypeError("Ticker of the current currency needs to be of str type.")
        
    # Check input type of curr1   
    if not isinstance(curr2, str):
        raise TypeError("Ticker of the desired currency needs to be of str type.")
        
    # Check input type of amt     
    if not isinstance(amt, (int, float)):
        raise TypeError("Amount of money to be converted needs to be a number.")
    
    curr_pair = curr1+curr2+'=X'
    
    data = yf.download(curr_pair, interval='1m', progress=False, show_errors=False)
    
    # Check for invalid ticker name
    if curr_pair in shared._ERRORS and 'symbol may be delisted' in shared._ERRORS[curr_pair] :
        raise NameError("You have entered an invalid foreign ticker! Try again.")
        
    conversion_rate = data['Close'].iloc[-1]
    converted_value = amt * conversion_rate
    
    return converted_value

src/fxtracker/fx_rate_lookup.py
# Authors: Sarah Abdelazim, Markus Nam, Crystal Geng, Lennon Au-Yeung
# Date: Jan 13, 2023

import datetime
import pandas as pd
# import pandas_datareader as pdr
import altair as alt
import numpy as np
import yfinance as yf


def fx_rate_lookup(curr, target_px):
    """
    Return the first date (reverse chronological order) on which the target price falling between day high and day low based on the availability of data.
    
    Parameters
    ----------
    curr : string 
        Ticker of the currency pair such as 'EURUSD'.
    target_px : float
        Target price for the lookup.
    
    Returns
    --------
    date: string
        The closest date in YYYY-MM-DD on which the target price falling between day high and day low.

    
    Examples
    --------
    >>> fx_rate_lookup('EURUSD', 1.072)
        '2023-01-10'
    """
    
    df = yf.download(curr + '=X', progress=False, show_errors=False)
    if len(df) == 0:
        raise Exception('No data found from data source. Check your ticker.')
    
    # A query to check if the input price is within the High and Low of any previous days.
    myquery = 'High >= ' + str(target_px) + 'and Low <= ' + str(target_px)
    mydates = df.query(myquery).index.format()
    
    # Found, return the latest one (closest to the query date)
    if len(mydates) > 0:
        return mydates[-1]
    else:
        raise Exception('Target price not found. Adjust your target price.')

src/fxtracker/price_trend_viz.py
# Authors: Sarah Abdelazim, Markus Nam, Crystal Geng, Lennon Au-Yeung
# Date: Jan 21, 2023

import datetime
import pandas as pd
import altair as alt
import numpy as np
import yfinance as yf
from yfinance import shared
import altair_viewer
alt.renderers.enable('mimetype')

def price_trend_viz(curr, start_date, end_date, option):
    """
    Visualizes trend of the exchange rate of a currency pair between the selected start date and end date.
    
    Parameters
    ----------
    curr : string 
        Ticker of the currency pair such as 'EURUSD'
    start_date : string 
        Start date of the selected period of time
    end_date : string
        End date of the selected period of time
    option: string
        A choice of option from ['Open', 'High', 'Low', 'Close'] 
    
    Returns
    --------
    Line plot that shows the trend of the exchange rate of a currency pair over the selected period of time
    
    Examples
    --------
    >>> price_trend_viz('EURUSD', '2018-12-31', '2022-12-31')
    """   
    
    # Check input type of curr
    if not isinstance(curr, str):
        raise TypeError("curr needs to be of str type.")
        
    # Check input type of start_date     
    if not isinstance(start_date, str):
        raise TypeError("start_date needs to be of str type.")
        
    # Check input type of end_date  
    if not isinstance(end_date, str):
        raise TypeError("end_date needs to be of str type.")
        
    if option not in ['Open', 'High', 'Low', 'Close']:
        raise Exception("Your option of plotting should be from 'Open', 'High', 'Low' or 'Close'")
     
    # Check if the end date entered is later than 2003-12-01
    if(datetime.datetime.strptime(end_date, "%Y-%m-%d") < datetime.datetime.strptime('2003-12-01', "%Y-%m-%d")):
        raise Exception("No data exists before 2003-12-01, please try again.")
        
    # Check if the start date entered is earlier than or equal to today 
    if(datetime.datetime.strptime(start_date, "%Y-%m-%d") > datetime.datetime.today()):
        raise Exception("You entered a start date later than today, please try again.") 
    
    curr_X = curr +'=X'   
        
    data = yf.download(curr_X, start_date, end_date, progress=False) 
    
    # Raise an exception if the data downloaded has emtpy content
    if len(data) == 0:
        raise Exception('No data found from data source. Check your ticker and date.') 

    
    trend_plot = alt.Chart(
        data.reset_index(), 
        title = f"Trend of Exchange Rate (Daily {option}) of {curr} from {start_date} to {end_date}"
    ).mark_line(
    ).encode(
        x = alt.X('Date', axis = alt.Axis(format = ("%Y-%m-%d"))),
        y = alt.Y(option, title = 'Exchange Rate', scale = alt.Scale(zero=False)),
        tooltip = alt.Tooltip(option, format = '.2%')
    ).properties(height=400, width=800
    ).configure_axis(
        labelFontSize=14,                                       
        titleFontSize=15     
    ).configure_title(
        fontSize=16)

    
    trend_plot.show()
    
    return trend_plot

src/fxtracker/fxtracker.py
# Authors: Sarah Abdelazim, Markus Nam, Crystal Geng, Lennon Au-Yeung
# Date: January, 2023

# import pandas_datareader as pdr
import yfinance as yf
import altair as alt
from yfinance import shared
import datetime
alt.renderers.enable('mimetype')


def fx_rate_lookup(curr, target_px):
    """
    Return the most recent date on which
    the target price happend.
    i.e. the target price was between day high
    and day low of the day.

    Parameters
    ----------
    curr : string
        Ticker of the currency pair such as 'EURUSD'.
    target_px : float
        Target price for the lookup.

    Returns
    --------
    date: string
        The most recent date in YYYY-MM-DD on which the target
        price happened. i.e. the target price was between day high
        and day low of that day.

    Examples
    --------
    >>> fx_rate_lookup('EURUSD', 1.072)
        '2023-01-10'
    """

    # check input type of curr
    if not isinstance(curr, str):
        raise TypeError("curr needs to be of str type.")

    # Check input type of target_px
    if not isinstance(target_px, (int, float)):
        raise TypeError("target_px needs to be a number.")

    df = yf.download(curr + '=X', progress=False, show_errors=False)
    if len(df) == 0:
        raise Exception('No data found from data source. Check your ticker.')

    # To check if the input price is within the High and Low of any previous
    # days.
    myquery = 'High >= ' + str(target_px) + 'and Low <= ' + str(target_px)
    mydates = df.query(myquery).index.format(
        formatter=lambda x: x.strftime('%Y-%m-%d'))

    # Found, return the latest one (closest to the query date)
    if len(mydates) > 0:
        return mydates[-1]
    else:
        raise Exception('Target price not found. Adjust your target price.')


def pl_trend_viz(curr, start_date, end_date, chart_type):
    """
    Visualizes trend of the profit and loss of a
    currency pair between the selected start date and end date.

    Parameters
    ----------
    curr : string
        Ticker of the currency pair such as 'EURUSD'
    start_date : string
        Start date of the selected period of time
    end_date : string
        End date of the selected period of time
    chart_type: string
        Type of visulization

    Returns
    --------
    Line plot that shows the trend of the profit and
    loss of a currency pair over the selected period of time

    Examples
    --------
    >>> pl_trend_viz('EURUSD', '2018-12-31', '2022-12-31','line')
    """
    shared._ERRORS = {}

    # check input type of curr
    if not isinstance(curr, str):
        raise TypeError("curr needs to be of str type.")

    curr_X = curr + '=X'

    # Check input type of start_date
    if not isinstance(start_date, str):
        raise TypeError("start_date needs to be of str type.")

    # Check input type of end_date
    if not isinstance(end_date, str):
        raise TypeError("end_date needs to be of str type.")

    # Check type of visulization
    if chart_type not in ['line', 'area']:
        raise ValueError(
            "Your option of plotting should be from 'line' or 'area'")

    # Assert end date is later than start date
    format = "%Y-%m-%d"
    if (datetime.datetime.strptime(end_date, format)
       < datetime.datetime.strptime(start_date, format)):
        raise ValueError(
            "The end date is earlier than the start date! Try again.")

    data = yf.download(curr_X, start_date, end_date)

    if (curr_X in shared._ERRORS and
            'symbol may be delisted' in shared._ERRORS[curr_X]):
        raise NameError(
            "You have entered an invalid foreign ticker! Try again.")

    drop = {'Open', 'High', 'Low', 'Adj Close', 'Volume'}
    data = data.drop(columns=drop)

    for i in range(1, len(data)):
        data.iloc[i, :] = round(
            (data.iloc[i, :] - data.iloc[0, :]) / data.iloc[0, :], 3)

    data.iloc[0, :] = round(
        (data.iloc[0, :] - data.iloc[0, :]) / data.iloc[0, :], 3)

    data = data.rename(columns={"Close": "Percentage Change"})

    title = 'Profit and loss trend over time for currency ' + curr

    if chart_type == 'line':
        chart = alt.Chart(data.reset_index(), title=title).mark_line().encode(
            x=alt.X('Date'),
            y=alt.Y('Percentage Change', axis=alt.Axis(format='%')),
            tooltip=alt.Tooltip('Percentage Change', format='.2%')
        )
    else:
        chart = alt.Chart(data.reset_index(), title=title).mark_area().encode(
            x=alt.X('Date'),
            y=alt.Y('Percentage Change', axis=alt.Axis(format='%')),
            tooltip=alt.Tooltip('Percentage Change', format='.2%')
        )

    # chart.show()

    return chart


def price_trend_viz(curr, start_date, end_date, option):
    """
    Visualizes trend of the exchange rate of a
    currency pair between the selected start date and end date.

    Parameters
    ----------
    curr : string
        Ticker of the currency pair such as 'EURUSD'
    start_date : string
        Start date of the selected period of time
    end_date : string
        End date of the selected period of time
    option: string
        A choice of option from ['Open', 'High', 'Low', 'Close']

    Returns
    --------
    Line plot that shows the trend of the exchange rate of
    a currency pair over the selected period of time

    Examples
    --------
    >>> price_trend_viz('EURUSD', '2018-12-31', '2022-12-31')
    """

    # Check input type of curr
    if not isinstance(curr, str):
        raise TypeError("curr needs to be of str type.")

    # Check input type of start_date
    if not isinstance(start_date, str):
        raise TypeError("start_date needs to be of str type.")

    # Check input type of end_date
    if not isinstance(end_date, str):
        raise TypeError("end_date needs to be of str type.")

    if option not in ['Open', 'High', 'Low', 'Close']:
        raise Exception(
            "Your option of plotting should be "
            "from 'Open', 'High', 'Low' or 'Close'")

    # Check if the end date entered is later than 2003-12-01
    if (datetime.datetime.strptime(end_date, "%Y-%m-%d")
            < datetime.datetime.strptime('2003-12-01', "%Y-%m-%d")):
        raise Exception("No data exists before 2003-12-01, please try again.")

    # Check if the start date entered is earlier than or equal to today
    if (datetime.datetime.strptime(start_date, "%Y-%m-%d")
            > datetime.datetime.today()):
        raise Exception(
            "You entered a start date later than today, please try again.")

    curr_X = curr + '=X'

    data = yf.download(curr_X, start_date, end_date, progress=False)

    # Raise an exception if the data downloaded has emtpy content
    if len(data) == 0:
        raise Exception(
            'No data found from data source. Check your ticker and date.')

    line_1 = f"Trend of Exchange Rate (Daily {option})"
    line_2 = f" of {curr} from {start_date} to {end_date}"

    trend_plot = alt.Chart(
        data.reset_index(),
        title=line_1+line_2
    ).mark_line(
    ).encode(
        x=alt.X('Date', axis=alt.Axis(format=("%Y-%m-%d"))),
        y=alt.Y(option, title='Exchange Rate', scale=alt.Scale(zero=False)),
        tooltip=alt.Tooltip(option, format='.2%')
    ).properties(height=400, width=600
                 ).configure_axis(
        labelFontSize=14,
        titleFontSize=15
    ).configure_title(
        fontSize=16)

    # trend_plot.show()

    return trend_plot


def fx_conversion(curr1, curr2, amt):
    """
    Converts a specific amount of money from current
    currency (curr1) to desired currency (curr2)

    Parameters
    ----------
    curr1 : string
        Ticker of the current currency such as 'EUR'
    curr2 : string
        Ticker of the desired currency to convert to such as 'USD'
    amt : int or float
        amount of money to be converted

    Returns
    --------
    Amount of value converted from currency 1 to currency 2

    Examples
    --------
    >>> fx_conversion('EUR', 'USD', 150.75)
    """
    shared._ERRORS = {}
    # Tests
    # Check input type of curr1
    if not isinstance(curr1, str):
        raise TypeError(
            "Ticker of the current currency needs to be of str type.")

    # Check input type of curr1
    if not isinstance(curr2, str):
        raise TypeError(
            "Ticker of the desired currency needs to be of str type.")

    # Check input type of amt
    if not isinstance(amt, (int, float)):
        raise TypeError(
            "Amount of money to be converted needs to be a number.")

    curr_pair = curr1 + curr2 + '=X'

    data = yf.download(
        curr_pair,
        interval='1m',
        progress=False,
        show_errors=False)

    # Check for invalid ticker name
    if (curr_pair in shared._ERRORS and
            'symbol may be delisted' in shared._ERRORS[curr_pair]):
        raise NameError(
            "You have entered an invalid foreign ticker! Try again.")

    conversion_rate = data['Close'].iloc[-1]
    converted_value = amt * conversion_rate

    return converted_value
